invokeai/frontend/web/src/features/nodes/util/graphBuilders/addHrfToGraph.ts

```diff
@@ -0,0 +1,210 @@
+import { RootState } from 'app/store/store';
+import { NonNullableGraph } from 'features/nodes/types/types';
+import {
+  DenoiseLatentsInvocation,
+  RescaleLatentsInvocation,
+  MetadataAccumulatorInvocation,
+  NoiseInvocation,
+  LatentsToImageInvocation,
+  Edge,
+} from 'services/api/types';
+import {
+  LATENTS_TO_IMAGE,
+  DENOISE_LATENTS,
+  NOISE,
+  MAIN_MODEL_LOADER,
+  METADATA_ACCUMULATOR,
+  LATENTS_TO_IMAGE_HRF,
+  DENOISE_LATENTS_HRF,
+  RESCALE_LATENTS,
+  NOISE_HRF,
+  VAE_LOADER,
+} from './constants';
+
+// Copy certain connections from previous DENOISE_LATENTS to new DENOISE_LATENTS_HRF.
+function copyConnectionsToDenoiseLatentsHrf(graph: NonNullableGraph): void {
+  const destinationFields = [
+    'vae',
+    'control',
+    'ip_adapter',
+    'metadata',
+    'unet',
+    'positive_conditioning',
+    'negative_conditioning',
+  ];
+  const newEdges: Edge[] = [];
+
+  // Loop through the existing edges connected to DENOISE_LATENTS
+  graph.edges.forEach((edge: Edge) => {
+    if (
+      edge.destination.node_id === DENOISE_LATENTS &&
+      destinationFields.includes(edge.destination.field)
+    ) {
+      // Add a similar connection to DENOISE_LATENTS_HRF
+      newEdges.push({
+        source: {
+          node_id: edge.source.node_id,
+          field: edge.source.field,
+        },
+        destination: {
+          node_id: DENOISE_LATENTS_HRF,
+          field: edge.destination.field,
+        },
+      });
+    }
+  });
+  graph.edges = graph.edges.concat(newEdges);
+}
+
+// Adds the high-res fix feature to the given graph.
+export const addHrfToGraph = (
+  state: RootState,
+  graph: NonNullableGraph
+): void => {
+  const { vae } = state.generation;
+  const isAutoVae = !vae;
+
+  // Pre-existing (original) graph nodes.
+  const originalDenoiseLatentsNode = graph.nodes[
+    DENOISE_LATENTS
+  ] as DenoiseLatentsInvocation;
+  const originalNoiseNode = graph.nodes[NOISE] as NoiseInvocation;
+  const originalLatentsToImageNode = graph.nodes[
+    LATENTS_TO_IMAGE
+  ] as LatentsToImageInvocation;
```

Comment:
Might be good to type these as optional eg `DenoiseLatentsInvocation | undefined`, that way we get accurate types when referring to them, and are forced to use optional chaining.

I fear we haven't done this throughout the rest of the graph functions.

Reply:
I might be misunderstanding our typescript works but graph.nodes cannot be undefined.

Reply:
Looks like I can set the node type to optionally undefined in the intialization of the variable. 

```
  // New node to convert latents to image.
  const latentsToImageHrfNode: LatentsToImageInvocation | undefined = originalLatentsToImageNode ? {
    type: originalLatentsToImageNode.type,
    id: LATENTS_TO_IMAGE_HRF,
    vae: originalLatentsToImageNode.vae,
    fp32: originalLatentsToImageNode.fp32,
    is_intermediate: originalLatentsToImageNode.is_intermediate,
  } : undefined;
  
  ```
  Is that what I should do?

Reply:
I meant this `const denoiseLatentsNode = graph.nodes[DENOISE_LATENTS] as DenoiseLatentsInvocation`

This tells TS that `denoiseLatentsNode` is never undefined, and is always a `DenoiseLatentsInvocation`, but that is not true. It could be undefined

The suggestion is to type it like this:
`const denoiseLatentsNode = graph.nodes[DENOISE_LATENTS] as DenoiseLatentsInvocation | undefined`

Then you're forced by TS - correctly - to check if `denoiseLatentsNode` is undefined before doing anything with it

Reply:
I'm a bit confused, I am defining the `denoiseLatentsHrfNode` above:
  // Define new nodes.
  // Denoise latents node to be run on upscaled latents.
  const denoiseLatentsHrfNode: DenoiseLatentsInvocation = {
    type: 'denoise_latents',
    id: DENOISE_LATENTS_HRF,
    is_intermediate: originalDenoiseLatentsNode?.is_intermediate,
    cfg_scale: originalDenoiseLatentsNode?.cfg_scale,
    scheduler: originalDenoiseLatentsNode?.scheduler,
    steps: originalDenoiseLatentsNode?.steps,
    denoising_start: 1 - state.generation.hrfStrength,
    denoising_end: 1,
  };
  
  so if we reach this step where it is added to the graph it should never be undefined correct?

Reply:
Not referring to the nodes you just created - I'm talking about the "original" nodes.

If you remove the casting (the `as DenoiseLatentsInvocation`), we will see that the type of `originalDenoiseLatentsNode` is an optional type - it could be `undefined`:
![image](https://github.com/invoke-ai/InvokeAI/assets/4822129/e8bbcb16-bb22-461d-ad3e-f07b07cea924)

If we cast it like `as DenoiseLatentsInvocation | undefined`, we are forced to handle the case where it is undefined.

Because the graph logic is getting pretty complicated, we need to be as conservative as possible with casting types with `as`. This exact situation has caused failed graph generation in the past.

The suggested change is:

```ts
  // Pre-existing (original) graph nodes.
  const originalDenoiseLatentsNode = graph.nodes[
    DENOISE_LATENTS
  ] as DenoiseLatentsInvocation | undefined;
  const originalNoiseNode = graph.nodes[NOISE] as NoiseInvocation | undefined;
  const originalLatentsToImageNode = graph.nodes[
    LATENTS_TO_IMAGE
  ] as LatentsToImageInvocation | undefined;
  ```

Then you'd need to handle those `undefined` edge cases

Reply:
ahhh gotcha, okay I added in these checks